File: AmigurumiMaker/app/src/main/java/com/example/amigurumimaker/domain/RowExpander.kt
package com.example.amigurumimaker.domain

import com.example.amigurumimaker.domain.model.RowDef

/**
 * Expands [RowDef] ranges into individual row definitions.
 * A RowDef with startRow=1, endRow=3 produces three RowDefs
 * with row numbers 1, 2, and 3, each carrying the same tokens.
 */
object RowExpander {

    fun expand(rowDef: RowDef): List<RowDef> {
        return (rowDef.startRow..rowDef.endRow).map { rowNumber ->
            RowDef(startRow = rowNumber, endRow = rowNumber, tokens = rowDef.tokens)
        }
    }
}

File: amigurumiMakerType/src/domain/RowExpander.ts
// src/domain/RowExpander.ts
// Port of Kotlin RowExpander — expands row definitions into individual rows

export interface RowDef {
  startRow: number;
  endRow: number;
  tokens: { type: string; count: number }[];
}

export interface ExpandedRow {
  startRow: number;
  tokens: { type: string; count: number }[];
}

/**
 * Expand a RowDef with a range of row numbers into individual ExpandedRows.
 * If startRow == endRow, returns a single row.
 */
export function expand(rowDef: RowDef): ExpandedRow[] {
  const rows: ExpandedRow[] = [];

  for (let row = rowDef.startRow; row <= rowDef.endRow; row++) {
    rows.push({
      startRow: row,
      tokens: rowDef.tokens.map((t) => ({ ...t })),
    });
  }

  return rows;
}
